Clean/Circuit/Circuit.lean

```diff
@@ -808,57 +705,47 @@ def assumptions (input : Vector (F p) 2) :=
   let ⟨ [x, y], _ ⟩ := input
   x.val < 256 ∧ y.val < 256
 
-def soundness_wrapped (inputs: Vector (F p) 2) (inputs_var: Vector (Expression (F p)) 2)
-    (h_inputs: (Provable.eval (F p) (α:=(fields (F p) 2)) inputs_var) = inputs)
-    (as: assumptions inputs)
-    (z: (F p))
-    (h: constraints_hold (add8_wrapped inputs_var (some z))) :
-    let z' := Provable.eval (F p) (output (add8_wrapped inputs_var (some z)))
-    spec inputs z'
-  := by
-  -- finish introductions
-  let ⟨ [x, y], _ ⟩ := inputs
-  let ⟨ [x_var, y_var], _ ⟩ := inputs_var
-  intro z'
-
-  -- characterize inputs
-  have h_inputs' : [x_var.eval, y_var.eval] = [x, y] := by injection h_inputs
-  injection h_inputs' with hx h_inputs'
-  injection h_inputs' with hy
-
-  -- characterize output, z' to equal (witness input) z, and replace in spec
-  have hz : z' = z := by rfl
-  rw [hz]
-
-  -- simplify constraints hypothesis
-  dsimp at h
-
-  -- h is just `subcircuit_soundness` of `Add8Full.circuit`
-  -- pass in the input values and a proof that they are correct
-  have h_inputs'' : vec [x_var.eval, y_var.eval, 0] = vec [x, y, 0] := by rw [hx, hy]
-  specialize h (vec [x, y, 0]) h_inputs''
-
-  -- satisfy `Add8Full.assumptions` by using our own assumptions
-  let ⟨ asx, asy ⟩ := as
-  have as': Add8Full.assumptions (vec [x, y, 0]) := ⟨asx, asy, by tauto⟩
-  specialize h as'
-
-  -- pass in output value and a (trivial) proof that it's correct
-  specialize h z rfl
-  guard_hyp h: Add8Full.circuit.spec (vec [x, y, 0]) z
-
-  -- unfold `Add8Full` statements to show what the hypothesis is in our context
-  dsimp [Add8Full.circuit, Add8Full.spec] at h
-  guard_hyp h: z.val = (x.val + y.val + (0 : F p).val) % 256
-
-  simp at h
-  exact h
-
-def circuit : FormalCircuit (F p) (fields (F p) 2) (field (F p)) (field (F p)) where
+def circuit : FormalCircuit (F p) (fields (F p) 2) (field (F p)) where
   main := add8_wrapped
   assumptions := assumptions
   spec := spec
-  soundness := soundness_wrapped
+  soundness := by
+    -- introductions
+    rintro env ⟨ inputs, _ ⟩ ⟨ inputs_var, _ ⟩ h_inputs as
+    let [x, y] := inputs
+    let [x_var, y_var] := inputs_var
+    intro h_holds z
+
+    -- characterize inputs
+    injection h_inputs with h_inputs
+    injection h_inputs with hx h_inputs
+    injection h_inputs with hy h_inputs
+    dsimp at hx hy
+
+    -- simplify constraints hypothesis
+    -- we know it must be just the `subcircuit_soundness` of `Add8Full.circuit`
+    -- so it has an implication form
+    have h_holds : _ → _ := h_holds
```

Comment:
another example of the same trick

Reply:
Same thing, just doing
```
dsimp [Adversarial.constraints_hold] at h_holds
```
works :slightly_smiling_face:

Reply:
wait what, am I crazy